File: packages/go/testkit/fixture_adapter.go
package testkit

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strings"

	aikit "github.com/Volpestyle/ai-kit/packages/go"
)

type FixtureKeyInput struct {
	Method   string
	Provider aikit.Provider
	Model    string
	Input    interface{}
}

type Fixture struct {
	Generate *aikit.GenerateOutput
	Stream   []aikit.StreamChunk
	Image    *aikit.ImageGenerateOutput
	Mesh     *aikit.MeshGenerateOutput
	Transcribe *aikit.TranscribeOutput
}

type FixtureCalls struct {
	Generate []aikit.GenerateInput
	Stream   []aikit.GenerateInput
	Image    []aikit.ImageGenerateInput
	Mesh     []aikit.MeshGenerateInput
	Transcribe []aikit.TranscribeInput
}

type FixtureAdapter struct {
	Provider  aikit.Provider
	Fixtures  map[string]Fixture
	Models    []aikit.ModelMetadata
	ChunkSize int
	KeyFunc   func(input FixtureKeyInput) string
	Calls     FixtureCalls
}

func (f *FixtureAdapter) ListModels(ctx context.Context) ([]aikit.ModelMetadata, error) {
	_ = ctx
	return f.Models, nil
}

func (f *FixtureAdapter) Generate(ctx context.Context, in aikit.GenerateInput) (aikit.GenerateOutput, error) {
	_ = ctx
	f.Calls.Generate = append(f.Calls.Generate, in)
	entry, err := f.lookupFixture("generate", in.Provider, in.Model, in)
	if err != nil {
		return aikit.GenerateOutput{}, err
	}
	if entry.Generate == nil {
		return aikit.GenerateOutput{}, fmt.Errorf("fixture for generate is missing (key: %s)", f.keyFor("generate", in.Provider, in.Model, in))
	}
	return *entry.Generate, nil
}

func (f *FixtureAdapter) GenerateImage(ctx context.Context, in aikit.ImageGenerateInput) (aikit.ImageGenerateOutput, error) {
	_ = ctx
	f.Calls.Image = append(f.Calls.Image, in)
	entry, err := f.lookupFixture("image", in.Provider, in.Model, in)
	if err != nil {
		return aikit.ImageGenerateOutput{}, err
	}
	if entry.Image == nil {
		return aikit.ImageGenerateOutput{}, fmt.Errorf("fixture for image is missing (key: %s)", f.keyFor("image", in.Provider, in.Model, in))
	}
	return *entry.Image, nil
}

func (f *FixtureAdapter) GenerateMesh(ctx context.Context, in aikit.MeshGenerateInput) (aikit.MeshGenerateOutput, error) {
	_ = ctx
	f.Calls.Mesh = append(f.Calls.Mesh, in)
	entry, err := f.lookupFixture("mesh", in.Provider, in.Model, in)
	if err != nil {
		return aikit.MeshGenerateOutput{}, err
	}
	if entry.Mesh == nil {
		return aikit.MeshGenerateOutput{}, fmt.Errorf("fixture for mesh is missing (key: %s)", f.keyFor("mesh", in.Provider, in.Model, in))
	}
	return *entry.Mesh, nil
}

func (f *FixtureAdapter) Transcribe(ctx context.Context, in aikit.TranscribeInput) (aikit.TranscribeOutput, error) {
	_ = ctx
	f.Calls.Transcribe = append(f.Calls.Transcribe, in)
	entry, err := f.lookupFixture("transcribe", in.Provider, in.Model, in)
	if err != nil {
		return aikit.TranscribeOutput{}, err
	}
	if entry.Transcribe == nil {
		return aikit.TranscribeOutput{}, fmt.Errorf("fixture for transcribe is missing (key: %s)", f.keyFor("transcribe", in.Provider, in.Model, in))
	}
	return *entry.Transcribe, nil
}

func (f *FixtureAdapter) Stream(ctx context.Context, in aikit.GenerateInput) (<-chan aikit.StreamChunk, error) {
	_ = ctx
	f.Calls.Stream = append(f.Calls.Stream, in)
	entry, err := f.lookupFixture("stream", in.Provider, in.Model, in)
	if err != nil {
		return nil, err
	}
	if entry.Stream != nil {
		return streamFromChunks(entry.Stream), nil
	}
	if entry.Generate == nil {
		return nil, fmt.Errorf("fixture for stream is missing (key: %s)", f.keyFor("stream", in.Provider, in.Model, in))
	}
	return streamFromChunks(buildStreamChunks(entry.Generate, f.chunkSize())), nil
}

func DefaultFixtureKey(input FixtureKeyInput) string {
	payload := struct {
		Method   string          `json:"method"`
		Provider aikit.Provider  `json:"provider"`
		Model    string          `json:"model"`
		Input    interface{}     `json:"input"`
	}{
		Method:   input.Method,
		Provider: input.Provider,
		Model:    input.Model,
		Input:    input.Input,
	}
	data, err := json.Marshal(payload)
	if err != nil {
		return fmt.Sprintf("%s:%s:%s:marshal_error", input.Method, input.Provider, input.Model)
	}
	hash := sha256.Sum256(data)
	return fmt.Sprintf("%s:%s:%s:%s", input.Method, input.Provider, input.Model, hex.EncodeToString(hash[:])[:12])
}

func GenerateKey(input aikit.GenerateInput) string {
	return DefaultFixtureKey(FixtureKeyInput{
		Method:   "generate",
		Provider: input.Provider,
		Model:    input.Model,
		Input:    input,
	})
}

func StreamKey(input aikit.GenerateInput) string {
	return DefaultFixtureKey(FixtureKeyInput{
		Method:   "stream",
		Provider: input.Provider,
		Model:    input.Model,
		Input:    input,
	})
}

func ImageKey(input aikit.ImageGenerateInput) string {
	return DefaultFixtureKey(FixtureKeyInput{
		Method:   "image",
		Provider: input.Provider,
		Model:    input.Model,
		Input:    input,
	})
}

func MeshKey(input aikit.MeshGenerateInput) string {
	return DefaultFixtureKey(FixtureKeyInput{
		Method:   "mesh",
		Provider: input.Provider,
		Model:    input.Model,
		Input:    input,
	})
}

func TranscribeKey(input aikit.TranscribeInput) string {
	return DefaultFixtureKey(FixtureKeyInput{
		Method:   "transcribe",
		Provider: input.Provider,
		Model:    input.Model,
		Input:    input,
	})
}

func buildStreamChunks(output *aikit.GenerateOutput, chunkSize int) []aikit.StreamChunk {
	if output == nil {
		return nil
	}
	chunks := make([]aikit.StreamChunk, 0)
	if output.Text != "" {
		for _, part := range chunkText(output.Text, chunkSize) {
			chunks = append(chunks, aikit.StreamChunk{Type: aikit.StreamChunkDelta, TextDelta: part})
		}
	}
	for idx := range output.ToolCalls {
		call := output.ToolCalls[idx]
		chunks = append(chunks, aikit.StreamChunk{Type: aikit.StreamChunkToolCall, Call: &call})
	}
	chunks = append(chunks, aikit.StreamChunk{
		Type:         aikit.StreamChunkMessageEnd,
		Usage:        output.Usage,
		FinishReason: output.FinishReason,
	})
	return chunks
}

func streamFromChunks(chunks []aikit.StreamChunk) <-chan aikit.StreamChunk {
	out := make(chan aikit.StreamChunk, len(chunks))
	go func() {
		defer close(out)
		for _, chunk := range chunks {
			out <- chunk
		}
	}()
	return out
}

func chunkText(text string, size int) []string {
	if text == "" {
		return nil
	}
	if size <= 0 {
		return []string{text}
	}
	chunks := make([]string, 0, (len(text)/size)+1)
	for start := 0; start < len(text); start += size {
		end := start + size
		if end > len(text) {
			end = len(text)
		}
		chunks = append(chunks, text[start:end])
	}
	return chunks
}

func (f *FixtureAdapter) keyFor(method string, provider aikit.Provider, model string, input interface{}) string {
	keyFn := f.KeyFunc
	if keyFn == nil {
		keyFn = DefaultFixtureKey
	}
	return keyFn(FixtureKeyInput{
		Method:   strings.TrimSpace(method),
		Provider: provider,
		Model:    model,
		Input:    input,
	})
}

func (f *FixtureAdapter) lookupFixture(method string, provider aikit.Provider, model string, input interface{}) (Fixture, error) {
	key := f.keyFor(method, provider, model, input)
	entry, ok := f.Fixtures[key]
	if !ok {
		return Fixture{}, fmt.Errorf("fixture not found (key: %s)", key)
	}
	return entry, nil
}

func (f *FixtureAdapter) chunkSize() int {
	if f.ChunkSize > 0 {
		return f.ChunkSize
	}
	return 24
}

File: packages/node/src/testing/fixtureAdapter.ts
import { createHash } from "crypto";
import { AiKitError } from "../core/errors.js";
import {
  ErrorKind,
  GenerateInput,
  GenerateOutput,
  ImageGenerateInput,
  ImageGenerateOutput,
  ListModelsParams,
  MeshGenerateInput,
  MeshGenerateOutput,
  ModelMetadata,
  Provider,
  StreamChunk,
  TranscribeInput,
  TranscribeOutput,
} from "../core/types.js";
import type { ProviderAdapter } from "../core/provider.js";

export type FixtureKeyInput =
  | { type: "generate"; input: GenerateInput }
  | { type: "stream"; input: GenerateInput }
  | { type: "image"; input: ImageGenerateInput }
  | { type: "mesh"; input: MeshGenerateInput }
  | { type: "transcribe"; input: TranscribeInput };

export type FixtureEntry = {
  generate?: GenerateOutput;
  stream?: StreamChunk[];
  image?: ImageGenerateOutput;
  mesh?: MeshGenerateOutput;
  transcribe?: TranscribeOutput;
};

export interface FixtureAdapterOptions {
  provider: Provider;
  fixtures: Record<string, FixtureEntry>;
  models?: ModelMetadata[];
  resolveKey?: (input: FixtureKeyInput) => string;
  defaultChunkSize?: number;
}

export interface FixtureCalls {
  generate: GenerateInput[];
  streamGenerate: GenerateInput[];
  generateImage: ImageGenerateInput[];
  generateMesh: MeshGenerateInput[];
  transcribe: TranscribeInput[];
}

export class FixtureAdapter implements ProviderAdapter {
  readonly provider: Provider;
  private readonly fixtures: Record<string, FixtureEntry>;
  private readonly models: ModelMetadata[];
  private readonly resolveKey: (input: FixtureKeyInput) => string;
  private readonly defaultChunkSize: number;
  readonly calls: FixtureCalls = {
    generate: [],
    streamGenerate: [],
    generateImage: [],
    generateMesh: [],
    transcribe: [],
  };

  constructor(options: FixtureAdapterOptions) {
    this.provider = options.provider;
    this.fixtures = options.fixtures;
    this.models = options.models ?? [];
    this.resolveKey = options.resolveKey ?? createFixtureKey;
    this.defaultChunkSize = options.defaultChunkSize ?? 24;
  }

  async listModels(params?: ListModelsParams): Promise<ModelMetadata[]> {
    if (params?.providers && !params.providers.includes(this.provider)) {
      return [];
    }
    return this.models;
  }

  async generate(input: GenerateInput): Promise<GenerateOutput> {
    this.calls.generate.push(input);
    const entry = this.requireFixture({ type: "generate", input });
    if (!entry.generate) {
      throw this.missingFixtureError("generate", input);
    }
    return entry.generate;
  }

  async *streamGenerate(input: GenerateInput): AsyncIterable<StreamChunk> {
    this.calls.streamGenerate.push(input);
    const entry = this.requireFixture({ type: "stream", input });
    if (entry.stream) {
      for (const chunk of entry.stream) {
        yield chunk;
      }
      return;
    }
    if (!entry.generate) {
      throw this.missingFixtureError("stream", input);
    }
    const chunks = buildStreamChunks(entry.generate, this.defaultChunkSize);
    for (const chunk of chunks) {
      yield chunk;
    }
  }

  async generateImage(input: ImageGenerateInput): Promise<ImageGenerateOutput> {
    this.calls.generateImage.push(input);
    const entry = this.requireFixture({ type: "image", input });
    if (!entry.image) {
      throw this.missingFixtureError("image", input);
    }
    return entry.image;
  }

  async generateMesh(input: MeshGenerateInput): Promise<MeshGenerateOutput> {
    this.calls.generateMesh.push(input);
    const entry = this.requireFixture({ type: "mesh", input });
    if (!entry.mesh) {
      throw this.missingFixtureError("mesh", input);
    }
    return entry.mesh;
  }

  async transcribe(input: TranscribeInput): Promise<TranscribeOutput> {
    this.calls.transcribe.push(input);
    const entry = this.requireFixture({ type: "transcribe", input });
    if (!entry.transcribe) {
      throw this.missingFixtureError("transcribe", input);
    }
    return entry.transcribe;
  }

  private requireFixture(input: FixtureKeyInput): FixtureEntry {
    const key = this.resolveKey(input);
    const entry = this.fixtures[key];
    if (!entry) {
      throw new AiKitError({
        kind: ErrorKind.Validation,
        message: `Fixture not found (key: ${key}).`,
        provider: this.provider,
      });
    }
    return entry;
  }

  private missingFixtureError(
    type: FixtureKeyInput["type"],
    input: GenerateInput | ImageGenerateInput | MeshGenerateInput | TranscribeInput,
  ): AiKitError {
    const key = this.resolveKey({ type, input } as FixtureKeyInput);
    return new AiKitError({
      kind: ErrorKind.Validation,
      message: `Fixture for ${type} is missing (key: ${key}).`,
      provider: this.provider,
    });
  }
}

export function createFixtureKey(input: FixtureKeyInput): string {
  const payload = stableStringify({
    type: input.type,
    provider: input.input.provider,
    model: input.input.model,
    input: scrubInput(input.input),
  });
  const hash = createHash("sha256").update(payload).digest("hex").slice(0, 12);
  return `${input.type}:${input.input.provider}:${input.input.model}:${hash}`;
}

export function buildStreamChunks(
  output: GenerateOutput,
  chunkSize: number,
): StreamChunk[] {
  const chunks: StreamChunk[] = [];
  if (output.text) {
    for (const part of chunkText(output.text, chunkSize)) {
      chunks.push({ type: "delta", textDelta: part });
    }
  }
  if (output.toolCalls) {
    for (const call of output.toolCalls) {
      chunks.push({ type: "tool_call", call });
    }
  }
  chunks.push({
    type: "message_end",
    usage: output.usage,
    finishReason: output.finishReason,
  });
  return chunks;
}

function chunkText(text: string, chunkSize: number): string[] {
  if (!text) {
    return [];
  }
  const size = chunkSize > 0 ? chunkSize : text.length;
  const chunks: string[] = [];
  for (let index = 0; index < text.length; index += size) {
    chunks.push(text.slice(index, index + size));
  }
  return chunks;
}

function scrubInput<T>(input: T): T {
  if (Array.isArray(input)) {
    return input.map((item) => scrubInput(item)) as T;
  }
  if (input && typeof input === "object") {
    const result: Record<string, unknown> = {};
    for (const key of Object.keys(input).sort()) {
      if (key === "signal") {
        continue;
      }
      const value = (input as Record<string, unknown>)[key];
      if (typeof value === "undefined" || typeof value === "function") {
        continue;
      }
      result[key] = scrubInput(value as object);
    }
    return result as T;
  }
  return input;
}

function stableStringify(value: unknown): string {
  return JSON.stringify(scrubInput(value));
}
